src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts

```diff
@@ -0,0 +1,120 @@
+import _ from 'lodash/fp'
+import { Fragment, useState } from 'react'
+import { h } from 'react-hyperscript-helpers'
+// @ts-expect-error
+import { ButtonPrimary, IdContainer } from 'src/components/common'
+import { centeredSpinner } from 'src/components/icons'
+import Modal from 'src/components/Modal'
+import { Ajax } from 'src/libs/ajax'
+import Events from 'src/libs/events'
+import { FormLabel } from 'src/libs/forms'
+import * as Nav from 'src/libs/nav'
+import { useCancellation, useOnMount } from 'src/libs/react-utils'
+import * as Utils from 'src/libs/utils'
+import { getExtension, getFileName } from 'src/pages/workspaces/workspace/analysis/file-utils'
+import { analysisNameInput, analysisNameValidator } from 'src/pages/workspaces/workspace/analysis/notebook-utils'
+import { analysisLauncherTabName } from 'src/pages/workspaces/workspace/analysis/runtime-common'
+import validate from 'validate.js'
+
+
+export const AnalysisDuplicator = ({ destroyOld = false, fromLauncher = false, printName, toolLabel, workspaceName, googleProject, workspaceId, namespace, bucketName, onDismiss, onSuccess }) => {
+  const [newName, setNewName] = useState('')
+  const [existingNames, setExistingNames] = useState([])
```

Comment:
When variables are initialized with an empty array, TS can't infer what the element type should be. But you can pass a type to `useState`.

```suggestion
  const [existingNames, setExistingNames] = useState<FileName[]>([])
```

This removes the need for `@ts-expect-error` on `setExistingNames(existingNames)`.

Reply:
This PR just aims to move AnalysisDuplicator into its own file. That way, the follow-up PR displays all actual type-related changes needed.

Reply:
Sounds good. In that case, I would have suggested leaving this as a JS file in this PR and converting it to TS in the follow up PR.